h/services/oauth_validator.py

```diff
@@ -348,6 +361,21 @@ def _find_refresh_token(self, value):
             .first()
         )
 
+    def _find_token(self, value):
+        """Retrieve a token without knowing which kind it is"""
+
+        if value is None:
+            return None
+
+        return (
+            self.session.query(models.Token)
+            .filter(
+                (models.Token.refresh_token == value) | (models.Token.value == value)
+            )
+            .order_by(models.Token.created.desc())
+            .first()
```

Comment:
Won't this throw if the token is not found? How is that handled? Since there is a unique constraint on both of these fields, there should only be zero or one result. In that case `one_or_none` may be more appropriate.

Reply:
This'll raise rather than returning `None` if there's no matching token. If I'm reading the code right it looks like the one place where this method is called (in `OAuthProviderService`) expects it to return `None` so I think it should be `.one_or_none()`.

Reply:
Righty ho. Looks like there should be a test around that too.

Reply:
I was following the pattern from `_find_refresh_token()`. Do we have the same problem there?

Reply:
I've added tests, but from the docs:

https://docs.sqlalchemy.org/en/13/orm/query.html#sqlalchemy.orm.query.Query.first

> Return the first result of this Query or None if the result doesn’t contain any row.

So I think we're safe either way.